File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // get lowest price, then highest price that occurs AFTER the lowest
        // use 2 pointers to compare prices in 1 loop
        int left = 0, right = 1, maxProfit = 0, profit = 0;
        while (right < prices.length) {
            if (prices[left] < prices[right]) {
                profit = prices[right] - prices[left];
                maxProfit = Math.max(maxProfit, profit);
            } else {
                left = right;
            }
            right++;
        }
        return maxProfit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # buy at l, expand window with r. if lower l, set l to that
        l=0
        res = 0
        for r in range(len(prices)):
            if prices[r] < prices[l]:
                l = r
            profit = prices[r] - prices[l]
            res = max(profit, res)
        return res
        